src/ray/core_worker/task_manager.cc

```diff
@@ -30,6 +30,65 @@ const int64_t kTaskFailureThrottlingThreshold = 50;
 // Throttle task failure logs to once this interval.
 const int64_t kTaskFailureLoggingFrequencyMillis = 5000;
 
+Status ObjectRefStream::TryReadNextItem(ObjectID *object_id_out) {
+  bool is_eof_set = last_ != -1;
+  if (is_eof_set && curr_ >= last_) {
+    RAY_LOG(DEBUG) << "ObjectRefStream of an id " << generator_id_
+                   << " has no more objects.";
+    *object_id_out = ObjectID::Nil();
+    return Status::ObjectRefStreamEoF("");
+  }
+
+  auto it = item_index_to_refs_.find(curr_);
+  if (it != item_index_to_refs_.end()) {
+    // If the current index has been written,
+    // return the object ref.
+    // The returned object ref will always have a ref count of 1.
+    // The caller of this API is supposed to remove the reference
+    // when the obtained object id goes out of scope.
+    *object_id_out = it->second;
+    curr_ += 1;
+    RAY_LOG_EVERY_MS(DEBUG, 10000) << "Get the next object id " << *object_id_out
+                                   << " generator id: " << generator_id_;
+  } else {
+    // If the current index hasn't been written, return nothing.
+    // The caller is supposed to retry.
+    RAY_LOG_EVERY_MS(DEBUG, 10000)
+        << "Object not available. Current index: " << curr_ << " last: " << last_
+        << " generator id: " << generator_id_;
+    *object_id_out = ObjectID::Nil();
+  }
+  return Status::OK();
+}
+
+bool ObjectRefStream::InsertToStream(const ObjectID &object_id, int64_t item_index) {
+  if (last_ != -1) {
+    RAY_CHECK(curr_ <= last_);
+  }
+
+  if (item_index < curr_) {
+    // Index is already used. Don't write it to the stream.
+    return false;
+  }
+
+  auto it = item_index_to_refs_.find(item_index);
+  if (it != item_index_to_refs_.end()) {
+    // It means the when a task is retried it returns a different object id
```

Comment:
hmm, how is it possible: object id is generated by us, for the same `item_index` we should always generate the same object id? The value can be different if the task is non-deterministic.